### PR TITLE
chore: Add Custom SwiftLint Rule for Jira Link in TODO Comments

### PR DESCRIPTION
**Pull request by agisilaos**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

This PR introduces a custom SwiftLint rule aimed at enhancing our code quality and task tracking. The new rule requires that every `TODO` comment in our codebase includes a specific Jira link, ensuring that all tasks are properly referenced and tracked.

### Changes
- Updated `.swiftlint.yml` to include a custom rule named `todo_requires_jira_link`.
- The rule checks for `TODO` comments and validates the presence of a Jira link following our specified format (`https://wearezeta.atlassian.net/browse/WPB-<ticket_number>`).
- Added appropriate warning messages for any `TODO` comments that don't comply with the rule.

### Impact
- Helps in maintaining a direct correlation between code comments (for pending tasks, enhancements, etc.) and corresponding Jira tickets.
- Streamlines our workflow by ensuring that all `TODO` items are easily traceable to our project management tool.
- Improves code documentation and project tracking.

### Testing
- Manually tested the new rule by running SwiftLint on various files to ensure it correctly identifies `TODO` comments both with and without the required Jira links.
- Verified that no other SwiftLint functionalities are affected by this change.

### Note
This rule will prompt a warning for any `TODO` comment that lacks a corresponding Jira link, aligning with our effort to maintain a robust and traceable code management system. All team members are encouraged to update existing `TODO` comments accordingly.

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by markusfassbender** [comment]
Thank you, just two things:
1. Now the places missing Jira links are failing 😁 maybe we need to add `//swiftlint:disable:this todo_requires_jira_link`
2. For me personally it's enough to write `[WPB-1234]` and I don't need a full link

[WPB-1234]: https://wearezeta.atlassian.net/browse/WPB-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by netbe** [comment]
@agisilaos maybe we should also check for FIXME

**Comment by agisilaos** [comment]
> @agisilaos maybe we should also check for FIXME

@netbe I will add it and I will let you know when it's ready!

**Comment by agisilaos** [comment]
> Thank you, just two things:
> 
> 1. Now the places missing Jira links are failing 😁 maybe we need to add `//swiftlint:disable:this todo_requires_jira_link`
> 2. For me personally it's enough to write `[WPB-1234]` and I don't need a full link

- Regarding the first point, I think it's good that they're failing since it's going to prompt us to fix them. If not, why should we have the rule in the first place?
- On the second point I was also thinking about it and wasn't sure really. I ended up going with the link route because it felt more straightforward. 

[WPB-1234]: https://wearezeta.atlassian.net/browse/WPB-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: .swiftlint.yml [hunk 1]
```diff
@@ -1,59 +1,67 @@
...
+  - legacy_constructor
+  - legacy_hashing
+  - no_fallthrough_only
+
+custom_rules:
+  todo_requires_jira_link:
+    name: "TODO: Requires Jira Link"
+    regex: '(TODO|FIX\s?ME|FIXME): (?!.*\\[WPB-\\d+\\])'
+    match_kinds: [comment]
+    message: "Do not add TODOs or FIX MEs without a JIRA reference. Include the JIRA ID like [WPB-680]."
```
Comment (markusfassbender) [inline .swiftlint.yml]: Maybe it's shorter and more encouraging (positive way to say it 😁):
`"Add JIRA reference to TODOs and FIX MEs like [WPB-680]."`

**Comment by markusfassbender** [comment]
Thanks @agisilaos !
> Regarding the first point, I think it's good that they're failing since it's going to prompt us to fix them. If not, why should we have the rule in the first place?

I think if we merge it to develop like this, all pull requests will fail on the check `SwiftLint`. Does it help or will everyone ignore it?
My suggestion is to add `//swiftlint:disable:this todo_requires_jira_link` to all existing problems and new ones will pop up then. When a developer comes by a disable comment, it should be added..

File: .swiftlint.yml (hunk 1 above)
Comment (agisilaos) [reply .swiftlint.yml]: Done!

**Comment by netbe** [comment]
> . Does it help or will everyone ignore it?

it will only fail on TODOs without jira ticket in your change, right @markusfassbender ?
If so it's ok, we'll clean a bit at the time

**Review by netbe — CHANGES_REQUESTED** [review]
in order to merge this, we would need that Swiftlint runs only on modified files on PR which is not the case.

I think we should first remove/solve the existing TODOs and merge this next

**Review by markusfassbender — APPROVED** [review]

File: wire-ios-sync-engine/Source/Synchronization/ZMOperationLoop.m [hunk 2]
```diff
@@ -98,7 +98,7 @@ - (void)tearDown;
     [ZMRequestAvailableNotification removeObserver:self];
     
     self.transportSession = nil;
-    ///TODO: 
+    // TODO:
```
Comment (netbe) [inline wire-ios-sync-engine/Source/Synchronization/ZMOperationLoop.m:101]: will this trigger it?

**Review by netbe — APPROVED** [review]

**Pull request merged**